File: mcp-rust-libs/packages/typescript/test/result-envelope.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import {
  ContractValidationError,
  ResultEnvelopeSchema,
  parseResultEnvelope,
  validateResultEnvelope,
} from "../src/generated/result-envelope.js";

type FixtureExpectation = {
  accepted: boolean;
  error_code?: string;
  error_path?: string;
};

const readJson = (url: URL): unknown =>
  JSON.parse(readFileSync(fileURLToPath(url), "utf8")) as unknown;

const expectationDocument = readJson(
  new URL("../../../contracts/examples/expectations.json", import.meta.url),
) as { fixtures: Record<string, FixtureExpectation> };

describe("ResultEnvelopeSchema", () => {
  for (const [fixture, expected] of Object.entries(expectationDocument.fixtures)) {
    it(`matches the canonical fixture ${fixture}`, () => {
      const input = readJson(new URL(`../../../contracts/examples/${fixture}`, import.meta.url));
      const stableError = validateResultEnvelope(input);
      expect(stableError === null).toBe(expected.accepted);
      expect(ResultEnvelopeSchema.safeParse(input).success).toBe(expected.accepted);
      if (!expected.accepted) {
        expect(stableError).toMatchObject({
          code: expected.error_code,
          path: expected.error_path,
        });
      }
    });
  }

  it("normalizes parse failures to stable code and JSON pointer", () => {
    expect(() =>
      parseResultEnvelope({
        ok: false,
        error: { code: "Bad", message: "invalid", retryable: false },
      }),
    ).toThrowError(ContractValidationError);
    try {
      parseResultEnvelope({
        ok: false,
        error: { code: "Bad", message: "invalid", retryable: false },
      });
    } catch (error) {
      expect(error).toMatchObject({ code: "pattern", path: "$/error/code" });
    }
  });
});

File: mcp-rust-libs/packages/dart/test/result_envelope_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:ore_mcp_contracts/ore_mcp_contracts.dart';
import 'package:test/test.dart';

void main() {
  final repositoryRoot = Directory.current.parent.parent;
  final examplesRoot = Directory('${repositoryRoot.path}/contracts/examples');
  final expectationDocument =
      jsonDecode(
            File('${examplesRoot.path}/expectations.json').readAsStringSync(),
          )
          as Map<String, Object?>;
  final fixtures = expectationDocument['fixtures']! as Map<String, Object?>;

  for (final entry in fixtures.entries) {
    final fixture = entry.key;
    final expected = entry.value! as Map<String, Object?>;
    test('matches the canonical fixture $fixture', () {
      final input = jsonDecode(
        File('${examplesRoot.path}/$fixture').readAsStringSync(),
      );
      ContractValidationException? failure;
      try {
        ResultEnvelope.fromJson(input);
      } on ContractValidationException catch (error) {
        failure = error;
      }
      expect(failure == null, expected['accepted']);
      if (expected['accepted'] == false) {
        expect(failure?.code, expected['error_code']);
        expect(failure?.path, expected['error_path']);
      }
    });
  }

  test('constructors enforce the same bounded invariants', () {
    expect(
      () => SafeError(
        code: 'bad.path',
        message: 'empty component',
        path: const [''],
        retryable: false,
      ),
      throwsA(
        isA<ContractValidationException>()
            .having((error) => error.code, 'code', 'min_length')
            .having((error) => error.path, 'path', r'$/error/path/0'),
      ),
    );
  });
}
